File: examples/javascript/agent_loop.mjs
// Full Tara agent loop in Node.js. No deps.
//
// Usage:
//   export TARA_API_KEY=tara_sk_...
//   node agent_loop.mjs "What's the weather in Dubai? And convert 100 USD to AED."

const TARA_API_BASE = "https://tara.tratok.com/api/v1";
const MAX_TURNS = 10;

// ---------- tool implementations ----------

function getWeather({ city }) {
  const data = {
    Dubai: { temp_c: 37, condition: "sunny" },
    "Abu Dhabi": { temp_c: 39, condition: "sunny" },
    London: { temp_c: 18, condition: "cloudy" },
  };
  return data[city] ?? { temp_c: 22, condition: "unknown", note: `no data for ${city}` };
}

function getCurrencyRate({ base, quote }) {
  const rates = {
    "USD/AED": 3.67,
    "AED/USD": 0.272,
    "USD/EUR": 0.92,
  };
  const k = `${base.toUpperCase()}/${quote.toUpperCase()}`;
  const rate = rates[k];
  if (!rate) return { error: `No rate for ${k}` };
  return { base: base.toUpperCase(), quote: quote.toUpperCase(), rate };
}

const tools = [
  {
    name: "get_weather",
    description: "Get current weather for a city. Returns temp_c and condition.",
    input_schema: {
      type: "object",
      properties: { city: { type: "string" } },
      required: ["city"],
    },
  },
  {
    name: "get_currency_rate",
    description: "Get the current exchange rate between two ISO-4217 currency codes.",
    input_schema: {
      type: "object",
      properties: {
        base: { type: "string" },
        quote: { type: "string" },
      },
      required: ["base", "quote"],
    },
  },
];

const handlers = {
  get_weather: getWeather,
  get_currency_rate: getCurrencyRate,
};

// ---------- HTTP call with simple retry ----------

async function postAgent({ messages, tools, maxAttempts = 5 }) {
  const apiKey = process.env.TARA_API_KEY;
  if (!apiKey) {
    console.error("Set TARA_API_KEY first.");
    process.exit(1);
  }
  const body = JSON.stringify({
    messages,
    tools,
    max_tokens: 1024,
    temperature: 0.7,
  });

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    const r = await fetch(`${TARA_API_BASE}/agent.php`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body,
    });

    if (r.status === 429) {
      const retryAfter = parseInt(r.headers.get("Retry-After") ?? "1", 10);
      await sleep((retryAfter + Math.random() * 0.5) * 1000);
      continue;
    }
    if (r.status >= 500) {
      await sleep((2 ** attempt + Math.random()) * 1000);
      continue;
    }
    if (!r.ok) {
      throw new Error(`HTTP ${r.status}: ${await r.text()}`);
    }
    return r.json();
  }
  throw new Error(`Tara /agent failed after ${maxAttempts} attempts`);
}

const sleep = (ms) => new Promise((res) => setTimeout(res, ms));

// ---------- the loop ----------

async function runAgent(userMessage) {
  const messages = [{ role: "user", content: userMessage }];

  for (let turn = 0; turn < MAX_TURNS; turn++) {
    console.log(`\n--- turn ${turn + 1} ---`);
    const response = await postAgent({ messages, tools });

    messages.push({ role: "assistant", content: response.content });

    for (const block of response.content) {
      if (block.type === "text") {
        console.log(`[text] ${block.text}`);
      } else if (block.type === "tool_use") {
        console.log(`[tool_use] ${block.name}(${JSON.stringify(block.input)})`);
      }
    }

    if (response.stop_reason === "end_turn") {
      return response.content
        .filter((b) => b.type === "text")
        .map((b) => b.text)
        .join("");
    }

    if (response.stop_reason === "tool_use") {
      const toolResults = [];
      for (const block of response.content) {
        if (block.type !== "tool_use") continue;
        const handler = handlers[block.name];
        if (!handler) {
          toolResults.push({
            type: "tool_result",
            tool_use_id: block.id,
            content: `Unknown tool: ${block.name}`,
            is_error: true,
          });
          continue;
        }
        try {
          const result = handler(block.input);
          console.log(`[tool_result] ${block.name} -> ${JSON.stringify(result)}`);
          toolResults.push({
            type: "tool_result",
            tool_use_id: block.id,
            content: JSON.stringify(result),
          });
        } catch (e) {
          toolResults.push({
            type: "tool_result",
            tool_use_id: block.id,
            content: `Tool raised: ${e.message}`,
            is_error: true,
          });
        }
      }
      messages.push({ role: "user", content: toolResults });
      continue;
    }

    if (response.stop_reason === "max_tokens") {
      return response.content
        .filter((b) => b.type === "text")
        .map((b) => b.text)
        .join("");
    }

    throw new Error(`Unexpected stop_reason: ${response.stop_reason}`);
  }

  throw new Error(`Agent loop exceeded ${MAX_TURNS} turns`);
}

const userInput =
  process.argv.slice(2).join(" ") ||
  "What's the weather in Dubai and Abu Dhabi? Also convert 100 USD to AED.";

const final = await runAgent(userInput);
console.log("\n=== final answer ===");
console.log(final);

File: examples/python/agent_loop.py
"""
Full Tara agent loop with tools.

This is the canonical pattern: declare tools, run a loop that alternates
between calling /agent and executing tool requests locally, terminating when
the model returns stop_reason="end_turn".

Usage:
    export TARA_API_KEY=tara_sk_...
    python agent_loop.py "What's the weather like in Dubai and Abu Dhabi?"
"""

from __future__ import annotations

import json
import os
import random
import sys
import time
from typing import Any, Callable

import requests

TARA_API_BASE = "https://tara.tratok.com/api/v1"
MAX_TURNS = 10  # safety cap


# ---------------------------------------------------------------------------
# Tool implementations — replace these with real ones in your app.
# ---------------------------------------------------------------------------

def tool_get_weather(city: str) -> dict[str, Any]:
    # Pretend weather. In a real app this would hit a weather API.
    fake = {
        "Dubai":     {"temp_c": 37, "condition": "sunny"},
        "Abu Dhabi": {"temp_c": 39, "condition": "sunny"},
        "London":    {"temp_c": 18, "condition": "cloudy"},
    }
    return fake.get(city, {"temp_c": 22, "condition": "unknown", "note": f"no data for {city}"})


def tool_get_currency_rate(base: str, quote: str) -> dict[str, Any]:
    # Pretend FX. In a real app this would hit an FX API.
    rates = {("USD", "AED"): 3.67, ("AED", "USD"): 0.272, ("USD", "EUR"): 0.92}
    rate = rates.get((base.upper(), quote.upper()))
    if rate is None:
        return {"error": f"No rate for {base}/{quote}"}
    return {"base": base.upper(), "quote": quote.upper(), "rate": rate}


TOOLS = [
    {
        "name": "get_weather",
        "description": "Get current weather for a city. Returns temp_c and condition.",
        "input_schema": {
            "type": "object",
            "properties": {
                "city": {"type": "string", "description": "City name in English."}
            },
            "required": ["city"],
        },
    },
    {
        "name": "get_currency_rate",
        "description": "Get the current exchange rate between two currency codes (ISO 4217).",
        "input_schema": {
            "type": "object",
            "properties": {
                "base": {"type": "string", "description": "Base currency, e.g. 'USD'"},
                "quote": {"type": "string", "description": "Quote currency, e.g. 'AED'"},
            },
            "required": ["base", "quote"],
        },
    },
]

TOOL_HANDLERS: dict[str, Callable[..., Any]] = {
    "get_weather": tool_get_weather,
    "get_currency_rate": tool_get_currency_rate,
}


# ---------------------------------------------------------------------------
# Tara client
# ---------------------------------------------------------------------------

def post_agent(messages: list[dict], tools: list[dict], max_attempts: int = 5) -> dict:
    api_key = os.environ.get("TARA_API_KEY")
    if not api_key:
        raise SystemExit("Set TARA_API_KEY in your environment first.")

    body = {
        "messages": messages,
        "tools": tools,
        "max_tokens": 1024,
        "temperature": 0.7,
    }

    last_err: Exception | None = None
    for attempt in range(max_attempts):
        try:
            r = requests.post(
                f"{TARA_API_BASE}/agent.php",
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
                json=body,
                timeout=60,
            )
            if r.status_code == 429:
                retry_after = int(r.headers.get("Retry-After", "1"))
                time.sleep(retry_after + random.uniform(0, 0.5))
                continue
            if 500 <= r.status_code < 600:
                wait = (2 ** attempt) + random.uniform(0, 1)
                time.sleep(wait)
                continue
            r.raise_for_status()
            return r.json()
        except requests.RequestException as e:
            last_err = e
            time.sleep((2 ** attempt) + random.uniform(0, 1))

    raise RuntimeError(f"Tara /agent call failed after {max_attempts} attempts: {last_err}")


# ---------------------------------------------------------------------------
# The loop
# ---------------------------------------------------------------------------

def run_agent(user_message: str) -> str:
    messages: list[dict] = [{"role": "user", "content": user_message}]

    for turn in range(MAX_TURNS):
        print(f"\n--- turn {turn + 1} ---")
        response = post_agent(messages=messages, tools=TOOLS)

        # Always append the assistant turn to history
        messages.append({"role": "assistant", "content": response["content"]})

        # Log what the model emitted
        for block in response["content"]:
            if block["type"] == "text":
                print(f"[text] {block['text']}")
            elif block["type"] == "tool_use":
                print(f"[tool_use] {block['name']}({json.dumps(block['input'])})")

        stop = response["stop_reason"]
        if stop == "end_turn":
            return "".join(b.get("text", "") for b in response["content"] if b["type"] == "text")

        if stop == "tool_use":
            tool_results = []
            for block in response["content"]:
                if block["type"] != "tool_use":
                    continue
                handler = TOOL_HANDLERS.get(block["name"])
                if handler is None:
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block["id"],
                        "content": f"Unknown tool: {block['name']}",
                        "is_error": True,
                    })
                    continue
                try:
                    result = handler(**block["input"])
                    print(f"[tool_result] {block['name']} -> {result}")
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block["id"],
                        "content": json.dumps(result),
                    })
                except Exception as e:
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block["id"],
                        "content": f"Tool raised: {e}",
                        "is_error": True,
                    })
            messages.append({"role": "user", "content": tool_results})
            continue

        if stop == "max_tokens":
            # The model hit its output cap. Could ask for more, but we stop.
            return "".join(b.get("text", "") for b in response["content"] if b["type"] == "text")

        raise RuntimeError(f"Unexpected stop_reason: {stop}")

    raise RuntimeError(f"Agent loop exceeded {MAX_TURNS} turns")


def main() -> None:
    user_message = (
        " ".join(sys.argv[1:])
        or "What's the weather like in Dubai and Abu Dhabi right now? "
        "Convert 100 USD to AED while you're at it."
    )
    final = run_agent(user_message)
    print("\n=== final answer ===")
    print(final)


if __name__ == "__main__":
    main()
